src/models/room.ts

```diff
@@ -1917,6 +1917,27 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
         // If any pending visibility change is waiting for this (older) event,
         this.applyPendingVisibilityEvents(event);
 
+        // Sliding Sync modifications:
+        // The proxy cannot guarantee every sent event will have a transaction_id field, so we need
+        // to check the event ID against the list of pending events if there is no transaction ID
+        // field. Only do this for events sent by us though as it's potentially expensive to loop
+        // the pending events map.
+        const txnId = event.getUnsigned().transaction_id;
+        if (!txnId && event.getSender() === this.myUserId) {
+            // check the txn map for a matching event ID
+            for (const tid in this.txnToEvent) {
+                const localEvent = this.txnToEvent[tid];
+                if (localEvent.getId() === event.getId()) {
+                    logger.debug("processLiveEvent: found sent event without txn ID: ", tid, event.getId());
+                    // update the unsigned field so we can re-use the same codepaths
+                    const unsigned = event.getUnsigned();
+                    unsigned.transaction_id = tid;
+                    event.setUnsigned(unsigned);
+                    break;
+                }
+            }
+        }
```

Comment:
This seems like it'll only work if the send request returns before the sliding sync response?

Reply:
Yes. The scenarios are:
 - /send returns then /sync returns : rely on the event ID as you cannot always rely on the txn ID existing.
 - /sync returns then /send returns : rely on txn ID to de-dupe as you don't know the event ID yet.

Given this is a proxy problem, and the proxy sits on the /sync hot-path, it's unlikely that /sync would return before /send. We can guard against it fully by checking the `txnToEvent` map when /send returns, then it will work in all scenarios.

Reply:
Given the packet switching nature of the network, and especially with the proxy likely running on a different server, you very much could experience higher latency on /send than /sync
Fixing this for all edges would be great